rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,315 @@
+found that data is returned as an object, otherwise this function will error.
+
+A metric will be emitted to indicate the lookup time.
+
+#### `search_rows`
+
+This function returns the results as an array of objects. If no row is found an
+empty array is returned. Multiple results just result in multiple rows in the
+array.
+
+A metric will be emitted to indicate the lookup time.
+
+#### Parameters
+
+*resource*
+
+The name of the resource to lookup. This must point to a resource specified
+in the config file eg `resources.csv_file`.
+
+*condition*
+
+`condition` is a single level, key/value object that specifies the fields and
+values to lookup. The fields must all match (AND) for the row to be returned.
+
+*case_sensitive*
+
+By default the search will be case insensitive, but this can be changed by
+passing `true` to this parameter.
+
+### Example config
+
+```toml
+[resources.csv]
+    type = "csv"
+    file = "/path/to/csv.csv"
+    delimiter = ","
+
+[sources.datadog_logs]
+    type = "datadog_logs"
+    address = "0.0.0.0:80"
+
+[transforms.simple_enrich]
+    type = "remap"
+    inputs = ["datadog_logs"]
+    source = '''
+        . = parse_json!(.message)
+
+        result, err = find_row(
+            resources.csv,
+            { "license_plate": .license }
+        )
+
+        if is_null(err) {
+            .first_name = result.first_name
+            .last_name = result.last_name
+        }
+   '''
+```
+
+## Internal Proposal
+
+We will need to add a new component type to Vector, call it `Resource`. On
+loading the Vector config these instances will be created and will load the
+data that they are pointing to.
+
+The entire data file will be loaded into memory, so all lookups will be
+performed in memory. A `Resource` will need to provide threadsafe,
+readonly access to the data that it loads.
+
+VRL will need to maintain the concept of `Resource`. This can be created as
+an additional element to the `vrl::Value` type. On compilation, the available
+resources can be added to the Variable type definitions. This ensures that
+during compilation the functions will access valid resources.
+
+### Indexing
+
+Although the first version is not going to be doing any indexing into the data
+it is worth bearing it in mind as we will most likely need to add this in due
+course.
+
+In order to perform the indexing VRL needs to know which fields to index. The
+criteria is being passed in as an object. If the type def for that object is
+known at runtime, we can extract the fields from this.
+
+If the type def isn't known we could
+
+- raise an error
+- perform the search unindexed
+- create the index at runtime - generally the shape of this object will not
+  change over time, so most likely the cost of creating the index would only
+  need to be paid on the first event.
+
+The approach we take here should be decided before the first release so we don't
+have to introduce a breaking change further down the line.
```

Comment:
Can we choose one and resolve the comment here?

Reply:
Yes, I vote "raise an error". 

There is no requirement for us to perform dynamic searches at present, so it would be a case of choosing the simplest option that works. Plus if we change it down the line it will be relaxing restrictions rather than increasing them, thus preserving backward compatibility.

Reply:
Works for me. Let's update with that then.